### PR TITLE
Generalise logic keeping track of state

### PR DESCRIPTION
**Pull request by AlexWaygood**
I was working on another PR for this project, and felt the need for another context manager similar to @JelleZijlstra's `allow_string_literals`. I realised that something like the pattern I propose in this PR might be a more extensible and less repetitive solution, so I've separated it out into its own PR.

### CODE REVIEW

File: pyi.py [hunk 1]
```diff
@@ -270,6 +270,27 @@ def _is_bad_TypedDict(node: ast.Call) -> bool:
...
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_inst: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        self.nesting -= 1
+
+    def __bool__(self) -> bool:
+        return bool(self.nesting)
```
Comment (Akuli) [inline pyi.py:291]: I think an operator overload is a bit too clever for this. How about `is_active()` or something?

File: pyi.py [hunk 2]
```diff
@@ -270,6 +270,27 @@ def _is_bad_TypedDict(node: ast.Call) -> bool:
     )
 
 
+@dataclass
+class AttributeManager:
```
Comment (AlexWaygood) [inline pyi.py:274]: Any ideas for a better name for this class?

Comment (Akuli) [inline pyi.py:274]: I don't like the name, but I can't think of a good name. How about `NestingCounter`?

File: pyi.py [hunk 3]
```diff
@@ -278,33 +299,10 @@ class PyiVisitor(ast.NodeVisitor):
...
-        """Determine whether we are inside a `def` statement"""
-        return bool(self._function_nesting)
 
-    @property
-    def in_class(self) -> bool:
-        """Determine whether we are inside a `class` statement"""
-        return bool(self._class_nesting)
+    string_literals_allowed: AttributeManager = AttributeManager()
+    in_function: AttributeManager = AttributeManager()
+    in_class: AttributeManager = AttributeManager()
```
Comment (Akuli) [inline pyi.py:305]: Do these really get created instance-specifically? I think you need to use `dataclasses.field`.

**Comment by JelleZijlstra** [comment]
I like the idea of generalizing this!

I'd prefer a more explicit notation though: `with self.in_class:` just sounds odd. How about we do `with self.in_class.enable():` and `if self.in_class.is_enabled():`?

File: pyi.py (hunk 3 above)
Comment (AlexWaygood) [reply pyi.py]: Oh, yup, you're right.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
I have no idea what the git diff algorithm is doing after my last two commits...

File: pyi.py [hunk 4]
```diff
@@ -271,40 +271,37 @@ def _is_bad_TypedDict(node: ast.Call) -> bool:
...
-    errors: list[Error] = field(default_factory=list)
-    # Mapping of all private TypeVars/ParamSpecs/TypeVarTuples to the nodes where they're defined
-    typevarlike_defs: dict[TypeVarInfo, ast.Assign] = field(default_factory=dict)
-    # Mapping of each name in the file to the no. of occurrences
-    all_name_occurrences: Counter[str] = field(default_factory=Counter)
-    _class_nesting: int = 0
-    _function_nesting: int = 0
-    _allow_string_literals: int = 0
+class NestingCounter:
+    """Context manager to help the PyiVisitor keep track of internal state"""
```
Comment (AlexWaygood) [inline pyi.py:275]:
```suggestion
    """Class to help the PyiVisitor keep track of internal state"""
```

**Review by Akuli — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Thanks both for the reviews!

**Pull request merged**